src/backend/cassandra/impl/QueryExecutor.h

```diff
@@ -0,0 +1,136 @@
+ * could leak easily. This version is slightly better but the overall design is
+ * flawed and should be reworked.
+ */
+template <
+    typename StatementType,
+    typename HandleType = Handle,
+    SomeRetryPolicy RetryPolicyType = ExponentialBackoffRetryPolicy>
+class AsyncQueryExecutor
+    : public std::enable_shared_from_this<
+          AsyncQueryExecutor<StatementType, HandleType, RetryPolicyType>>
+{
+    using FutureWithCallbackType = typename HandleType::FutureWithCallbackType;
+    using CallbackType =
+        std::function<void(typename HandleType::ResultOrErrorType)>;
+
+    clio::Logger log_{"Backend"};
+
+    StatementType data_;
+    RetryPolicyType retryPolicy_;
+    CallbackType onComplete_;
+
+    // does not exist during initial construction, hence optional
+    std::optional<FutureWithCallbackType> future_;
+
+public:
+    /**
+     * @brief Create a new instance of the AsyncQueryExecutor and execute it.
+     */
+    static void
+    run(boost::asio::io_context& ioc,
+        HandleType const& handle,
+        StatementType data,
+        CallbackType&& onComplete)
+    {
+        // this is a helper that allows us to use std::make_shared below
+        struct EnableMakeShared : public AsyncQueryExecutor<
+                                      StatementType,
+                                      HandleType,
+                                      RetryPolicyType>
+        {
+            EnableMakeShared(
+                boost::asio::io_context& ioc,
+                StatementType&& data,
+                CallbackType&& onComplete)
+                : AsyncQueryExecutor(
+                      ioc,
+                      std::move(data),
+                      std::move(onComplete))
+            {
+            }
+        };
+
+        auto ptr = std::make_shared<EnableMakeShared>(
+            ioc, std::move(data), std::move(onComplete));
+        ptr->execute(handle);
+    }
+
+private:
+    AsyncQueryExecutor(
+        boost::asio::io_context& ioc,
+        StatementType&& data,
+        CallbackType&& onComplete)
+        : data_{std::move(data)}
+        , retryPolicy_{ioc}
+        , onComplete_{std::move(onComplete)}
+    {
+    }
+
+    void
+    execute(HandleType const& handle)
+    {
+        auto self = this->shared_from_this();
+
+        // lifetime is extended by capturing self ptr
+        auto handler = [this, &handle, self](auto&& res) mutable {
+            if (res)
+            {
+                onComplete_(std::move(res));
+            }
+            else
+            {
+                if (retryPolicy_.shouldRetry(res.error()))
+                    retryPolicy_.retry(
+                        [self, &handle]() { self->execute(handle); });
+                else
+                    onComplete_(std::move(res));  // report error
+            }
+        };
+        future_.emplace(handle.asyncExecute(data_, std::move(handler)));
```

Comment:
Do we really need this future_? seems like we don't use it after creation.

Reply:
This future is the second component that keeps this whole thing alive - the future is holding onto the callback that captures `self` to extend lifetime. If we don't store this future then our callback would be released leading to a crash once it's called by the cassandra-driver.

Reply:
Yea~ Driver API gets the callback's ptr but not the object.